### DISCUSSION TITLE
Solver choice, calculated Psi

### DISCUSSION THREAD

**Discussion by rugfri** [PyEllips/pyElli#158] [General]
Dear Pyelli's

I've got some data measured with a Sentech ellipsometer and I was able to fit.
On starting I get the following warning: 

UserWarning: Solver2x2 can't handle active media (n > 0 and k < 0).
                    Check if all materials are defined correctly or switch to Solver4x4 instead.

Which I didn't get when fitting the data from the same sample using a Woollam instrument instead.

Is it because k is very small and thus  during evaluation sometimes k < 0?

Furthermore, in the trace of the calculated Psi (continuous cyan line in the figure attached), I can see something resembling a singularity. 
![fit_crop](https://github.com/PyEllips/pyElli/assets/34413846/ae5aa942-d304-4979-8c2e-7c7bea2d8c04)
Where it does it come from?

Thanks!

**Comment by MarJMue** [comment]
Hi,

> UserWarning: Solver2x2 can't handle active media (n > 0 and k < 0).
>                     Check if all materials are defined correctly or switch to Solver4x4 instead.
> 
> Which I didn't get when fitting the data from the same sample using a Woollam instrument instead.

this warning was implemented fairly recently. Maybe thats why you did not encounter it before.
I implemented it because the solver does introduce phase flips for delta, if the media becomes active. So its should be more of a hint to check if everything seems plausible.

But the check only handles the dispersions of the model and should only be caused by a measurement, if a fit leads to a model, which triggers the warning.

> Is it because k is very small and thus  during evaluation sometimes k < 0?

Exactly this has happened to me, while interpolating tabular values. But that case should be fixed. I will try and see if i did any mistakes.

> Furthermore, in the trace of the calculated Psi (continuous cyan line in the figure attached), I can see something resembling a singularity. 

Without remembering the samples you used, it looks to me like a Sellmeier oscillator moved into your fitting range. But I can't be certain, without knowing how your model is built.

**Comment by andreaserbe** [comment]
One question I would ask myself: Does the Delta look the same (or at least similar) after measuring with Sentech and with Woollam? There may be different ways to define the range of Delta which can influence the fits. (But then maybe you would not be able to fit at all?). 
(And btw @Pyelli's - great job!).

**Reply by domna** [reply]
True, there are differences but as you already stated they would show a stronger difference and we would not be able to fit them so nicely. There normally is a shift of ranges for delta (delta typically is in the range [0, 180], [-180, 180] or [0, 360] degrees)

(Thanks 🙃)

**Reply by rugfri** [reply]
yes,  Woollam "delta" is in the [0,180], while Sentech in [0,360] but pyElli already handles that nicely :-)

**Comment by domna** [comment]
Hey @rugfri,

thanks for reaching out. I think this can happen whenever any of your layers get a negative k during the fitting. This may mean that your model is running out of physical plausible values during the fit. I have the feeling that the spike you see in your data is originating from this unphysical value but it's hard to pinpoint (as @MarJMue already pointed out) without knowing your actual models and parameters. I don't think that this is an oscillator which moved into the data range, because I think we would also see a peak in delta then (which I assume is the red curve).

One way to debug this would be to plot all permittivities of all layers with the fitted parameters and see where k is negative. It would also help us if you can provide the models you used with the parameters you fitted them to.

**Comment by rugfri** [comment]
@domna 
thanks
I created  a notebook in NOMAD, /uploads/wvase-testing.../solver_test.ipynb.
The parameters are updated with the ones I get from the fit.

I was also wondering if such behavior is the sign of something in the sample that my model does not fully captures and the fitting algorithm is somehow/necessarily trying to reproduce using the parameters. Well, sounds speculative. However the material is anisotropic in nature, but given its domain-like microstructure I tend to think that such effect should be very small, at least.

**Reply by domna** [reply]
Thank you. Could you also provide the parameters _after_ fitting? Then I could investigate the model and see if something went wrong somewhere. I think it should also show you the warning when you evaluate the model once with this parameter set.

**Reply by rugfri** [reply]
Hi @domna, I updated the notebook.
Thanks

**Reply by domna** [reply]
Strangely, I don't get see your "singularity" in the psi values with these parameter set. Was it showing something for you?
One thing I've already seen is that the broadening for the oscillator is very small (~1e-7). This is rather untypical, because such small resonances are not very common, or do you expect such small resonances from your material?

Here, you see the dispersion with the sharp resonance:
![Bildschirmfoto 2024-01-03 um 10 30 38](https://github.com/PyEllips/pyElli/assets/1014422/77e62827-fead-4e8a-b353-27345a7ae400)

**Reply by rugfri** [reply]
Strange, indeed, since the message appears I would say at the beginning of the fit.
I do expect k \approx 0, or at least a the sensitivity limit and resonance as well.
I tend to use the broadening as free parameter, to have a sort of 'flag' to spot "bad" samples. 
Fixing \Gamma to 0, still produces the warning.
I wonder if I should take this as a hint for an incomplete model.
Thanks anyway!

**Reply by domna** [reply]
I understand. It's just happening during the fit. That's tricky to avoid, but you can try to restrict your model parameters to physical meaningful values. I suspect that the Lorentz is moving somewhere out of plausible values which produces negative k's.

$\Gamma = 0$ are also somewhat unphysical values for the Lorentz oscillator. I think it would be helpful to set it to a minimum of 1e-3 or 1e-4. Generally you could try to restrict the parameter ranges such that it does not run out of any unphysical values. I think for your params setting $E > 0$ and $\Gamma > 0$ should help.
